Address slither reentrancy findings

cliffhall opened the pull request:
* In slither.db.json
  - updated via slither's triage-mode, but something's bonkers about slither and sometimes things stay ignored, sometimes they don't. Also just running it in normal mode several times will give different results. Therefore, I have no confidence that we'll ever get it to totally shut up, and we should just consider it helpful guide, and not like code coverage output which can eventually reach 100%

* In BosonRouter.sol,
  - removed a lot of whitespace
  - added virtual keyword to requestCreateOrderETHETH method, so it can be overriden by MockBosonRouter.sol
  - added nonReentrant modifier to methods:
    - requestCreateOrderTKNTKNWithPermit
    - requestCreateOrderTKNTKNWithPermitConditional
    - requestCreateOrderETHTKNWithPermit
    - requestCreateOrderETHTKNWithPermitConditional
    - requestCreateOrderTKNETH
    - requestCreateOrderTKNETHConditional
  - removed nonReentrant modifier from requestCreateOrderTKNETHInternal method, which is internal and the modifier has already been invoked on all callers.

* In Cashier.sol,
  - added nonReentrant modifier to methods:
    - onERC721Transfer
    - onERC1155Transfer

* In DAITokenWrapper.sol,
  - imported and extended ReentrancyGuard
  - added nonReentrant modifier to methods:
    - permit

* In ERC1155ERC721.sol,
  - imported and extended ReentrancyGuard
  - added nonReentrant modifier to methods:
    - safeTransferFrom
    - _transferFrom

* In MockBosonRouter.sol,
  - removed all imports except BosonRouter
  - extended BosonRouter
  - added constructor that passes its args to BosonRouter
  - removed all functions except
    - requestCreateOrderETHETH (which sends an invalid payment type to requestCreateOrder for a specific test)
    - transferFromAndAddEscrowTest

* In VoucherKernel.sol,
  - imported and extended ReentrancyGuard
  - added nonReentrant modifier to methods:
    - fillOrder
    - cancelOrFaultVoucherSet

This fixes #177

cliffhall committed:
Address slither reentrancy findings

* In slither.db.json
  - updated via slither's triage-mode, but something's bonkers about slither and sometimes things stay ignored, sometimes they don't. Also just running it in normal mode several times will give different results. Therefore, I have no confidence that we'll ever get it to totally shut up, and we should just consider it helpful guide, and not like code coverage output which can eventually reach 100%

* In BosonRouter.sol,
  - removed a lot of whitespace
  - added virtual keyword to requestCreateOrderETHETH method, so it can be overriden by MockBosonRouter.sol
  - added nonReentrant modifier to methods:
    - requestCreateOrderTKNTKNWithPermit
    - requestCreateOrderTKNTKNWithPermitConditional
    - requestCreateOrderETHTKNWithPermit
    - requestCreateOrderETHTKNWithPermitConditional
    - requestCreateOrderTKNETH
    - requestCreateOrderTKNETHConditional
  - removed nonReentrant modifier from requestCreateOrderTKNETHInternal method, which is internal and the modifier has already been invoked on all callers.

* In Cashier.sol,
  - added nonReentrant modifier to methods:
    - onERC721Transfer
    - onERC1155Transfer

* In DAITokenWrapper.sol,
  - imported and extended ReentrancyGuard
  - added nonReentrant modifier to methods:
    - permit

* In ERC1155ERC721.sol,
  - imported and extended ReentrancyGuard
  - added nonReentrant modifier to methods:
    - safeTransferFrom
    - _transferFrom

* In MockBosonRouter.sol,
  - removed all imports except BosonRouter
  - extended BosonRouter
  - added constructor that passes its args to BosonRouter
  - removed all functions except
    - requestCreateOrderETHETH (which sends an invalid payment type to requestCreateOrder for a specific test)
    - transferFromAndAddEscrowTest

* In VoucherKernel.sol,
  - imported and extended ReentrancyGuard
  - added nonReentrant modifier to methods:
    - fillOrder
    - cancelOrFaultVoucherSet

This fixes #177

hswopeams reviewed (changes requested):
Looks good for the most part. I have one request for change and a few questions. See inline.

cliffhall committed:
Add solhint comment

* In MockBosonRouter.sol,
  - in constructor, added a solhint-disable-next-line no-empty-blocks comment before the empty constuctor block

cliffhall committed:
Add nonReentrant modifier

* In VoucherKernel.sol,
  - in createTokenSupplyId method,
    - added nonReentrant modifier

hswopeams reviewed (dismissed):
Approving assuming the build issue will get fixed.

cliffhall committed:
Tweak formatting in MockBosonRouter.sol to fix creaky build.

cliffhall committed:
Move solhint comment in MockBosonRouter.sol

cliffhall committed:
Let prettier have a crack at fixing MockBosonRouter.sol

bidyut-arianelabs reviewed (approved):
Looks good to me 👍

cliffhall committed:
Merge branch 'main' into ignore-some-slither-reentrancy-findings
